doc/specifications/proposed/replicator/architecture.rst

```diff
@@ -0,0 +1,142 @@
+
+A simple auto-increment in the write-master might be the simplest solution. You
+should take care for possible overflows, though. The revision might increase
+quite fast in scenarios with a lot of updates.
+
+Idempotent Updates
+------------------
+
+In some cases it is helpful if all updates are idempotent, thus applying the
+same update multiple times does no harm.
+
+In this case the last revision on the target system can be reset without
+resetting the full storage. This can be helpful for graceful recovery of the
+target system.
+
+Compaction
+----------
+
+One problem with this approach is that if the replication target had gone away
+every revision of every document will be re-transmitted in order. This can lead
+to quite long initializations.
+
+An option commonly used is to compact the database, thus removing deleted
+objects and older revisions so that they won't be re-transmitted. One must pay
+attention to not affect data integrity when doing this.
+
+Implementation
+==============
+
+The process may stop and continue at any point. If a replication fails at any
+point this is no problem if the process had been implemented correctly.
+
+Thus the updater / replicator can be implemented as a daemon, cronjob or a
+script triggered manually from time to time.
+
+The updater / replicator generally can run anywhere, on the source system, on
+the target system or on a dedicated system. The only respective drawback might
+be additional network latency.
+
+The updater / replicator should not do any data processing itself. This way it
+can be a very simple piece of software.
+
+Protocol
+--------
+
+You basically can implement this using any protocol you want. We implemented
+this using something like Json-RPC and XML-RPC. But you might come up with
+anything else – the messages are fairly simple. The most complex task is
+properly encoding the updates which need to be transfered.
+
+Displaying Replication Status
+=============================
+
+Since the process of storing updates is not necessarily immediate anymore one
+might want to give the user some impression on the replication status. To do
+this one can store information about the last requested updates by the
+replicator / updater on the source system. Imagine the following update
+stream, while the updater requests 5 updates starting at revision 3::
+
+    0 1 2 3 4 5 6 7 8 9
+          ^--------
+           ?since=2&limit=5
+
+We now know the following:
+
+1) The updates 0, 1 & 2 are definitely stored on the target system
+
+2) The updates 3, 4, 5, 6 & 7 are currently in transfer. Their state on the
+   target system is unclear until the next request for updates.
+
+3) The updates 8 & 9 are pending and not yet transferred.
+
+Given this information one could display the use the replication state of each
+document (green, yellow, red). The only additional information to store is last
+requested revision and the last transferred revision resulting from the
+requested limit.
```

Comment:
I think the document should go in greater detail to describe debugging and verification protocols.

In my own experience with ezcontentstaging, which has been used in real-life for quite a while, out-of-sync situations do happen, despite the best efforts in coding, and having good troubleshooting/fixing tools is a must.